### ISSUE TITLE
Automatically delete DB entries when plugins are uninstalled, add new reset functionality

### ISSUE THREAD

**Issue by pitaj** [NodeBB/NodeBB#2614]
NodeBB should probably automatically delete DB entries for plugins before they are removed and reset. 

First of all, there would be a new `RESET` button (one button for each plugin, like `DEACTIVATE` and `UNINSTALL` buttons) in the ACP 

This could be done in two ways:

 1. Plugins register their DB fields in plugin.json (under something like `db-fields`). These field are automatically deleted or cleared when the plugin is uninstalled / reset
 2. The `database` module records all of the different fields the plugin uses somehow, and then that list is used to reset the plugins

I think the first way is preferable, as it would be easier to implement and would allow for more developer flexibility.

Sample plugin.json: 

```json
{
   ...
   "db-fields": [
    "plugin-calendar:settings",
    "plugin-calendar:events",
    "plugin-calendar:event[0-9]+"
  ]
}
```

The fields array could allow for regular expressions to match various different DB fields (like above).

**Comment by julianlam** [comment]
The plugin system gives developers the power to do anything nodebb itself can. This is part of the reason they're as flexible as they are. There is no sandbox. 

By that same token it does not make sense to force developers to register their database values in order for NodeBB to clean them up. 

The situation gets more frustrating when you throw in database schema changes (they happen even in plugins!).

Naturally, if a plugin author were in need of the ability to purge plugin related data, it would be the same amount of effort to register a listener to a new `static:plugin.reset` hook, and handle the cleanup itself. 

Perhaps a best practices guide for plugin authors is a good course of action, in addition to adding the new hook. 

**tl;dr** - people are lazy and won't register their custom data, I'd rather place the onus on the plugin dev to handle it.

**Comment by julianlam** [comment]
I don't mean to sound offensive (I tend to, at times)! I just want to communicate how monumental a task this would be, in addition to the ongoing maintenance of the code to handle the field deletion. 

Also some plugins (especially early ones) add their configs straight to the `config` hash, many add fields to topic hashes... some other ones save physical files to disk... :smile:

**Comment by barisusakli** [comment]
Added action:plugin.install and action:plugin.uninstall
https://github.com/NodeBB/NodeBB/commit/3e666adfea44bca37a786e6d066e93aedc9359b9

In the plugin can remove the data from db using the uninstall hook. Updated the rss plugin to do just that. 

https://github.com/barisusakli/nodebb-plugin-rss/commit/8a457face3e37ac94c8a14159295fd0aca4b2754

**Comment by pitaj** [comment]
Awesome, @barisusakli
